### DISCUSSION TITLE
Improve support for Shared Projects

### DISCUSSION THREAD

**Discussion by HavenDV** [unoplatform/uno#8019] [Ideas]
Does anyone know if Microsoft has plans to improve support for Shared Projects and add proper recognition of wildcard expressions to them? So that they behave according to sdk like projects in terms of including files. For Uno projects, this would come in handy.
Why the question - perhaps there is an issue somewhere that I could not find, for which I and others can vote and speed up the process.
I also found it: https://developercommunity2.visualstudio.com/t/Add-shared-projects-item-groups-globbing/588022?from=email&space=8&q=Shared+Project&entry=suggestion
Perhaps it's worth trying again? Now there are more of us. There are more arguments too.
I know you've added a Globbed ItemGroup to the uno template shared project, but it's still quite inconvenient when adding new files or copying through VS UI. Also, after the last update of VS Preview, WinUI project started giving me the following error message with Globbed ItemGroup: 
`error NETSDK1022: Duplicate 'Compile' items were included.`
However, this only applies to files that are included implicitly and via _Globbed_Compile

**Comment by jeromelaban** [comment]
Microsoft is in a conflicting position when dealing with Shared project. It's deprecated but in practice definitely is not, as it's a great tool for dealing with project incompatibilities.

Now with the `error NETSDK1022: Duplicate 'Compile' items were included.` error, this is different. We're not seeing this error in builds, where do you see it?

**Reply by HavenDV** [reply]
> Now with the `error NETSDK1022: Duplicate 'Compile' items were included.` error, this is different. We're not seeing this error in builds, where do you see it?

Through some tests, I realized that this manifests itself if there are two or more Shared projects that are used in one WinUI project and have the following lines in `.projitems` file:
```xml
<ItemGroup>
    <!--
		 Add files present in the shared project folder, excluding the ones
		 explicitly included in the projitem, so files can be added from vscode
		 without manipulating the projitem file.
	-->
    <_Globbled_Page Include="$(MSBuildThisFileDirectory)**/*.xaml" Exclude="@(Page);@(ApplicationDefinition)">
      <SubType>Designer</SubType>
      <Generator>MSBuild:Compile</Generator>
    </_Globbled_Page>
    <Page Include="@(_Globbled_Page)">
      <SubType>Designer</SubType>
    </Page>
    <_Globbed_Compile Include="$(MSBuildThisFileDirectory)**/*.xaml.cs" Exclude="@(Compile)">
      <DependentUpon>%(Filename)</DependentUpon>
    </_Globbed_Compile>
    <_Globbed_Compile Include="$(MSBuildThisFileDirectory)**/*.cs" Exclude="@(Compile);@(_Globbed_Compile)" />
    <Compile Include="@(_Globbed_Compile)" />
    <_Globbed_PRIResource Include="$(MSBuildThisFileDirectory)**/*.resw" Exclude="@(PRIResource)" />
    <PRIResource Include="@(_Globbed_PRIResource)" />
    <_Globbed_Content Include="$(MSBuildThisFileDirectory)Assets/**/*.*" Exclude="@(Content)" />
    <Content Include="@(_Globbed_Content)" />
  </ItemGroup>
```
I tried to add `;@(_Globbed_Compile)` to first `_Globbed_Compile` to the file that is imported last, but it had no effect

P.S. The solution to this problem is to simply rename _Globbed_ to _Globbed2/3/4/5_

**Reply by jeromelaban** [reply]
Thanks for the investigation! Would you be able to contribute a change ?

**Reply by HavenDV** [reply]
This is not required. The original Uno template contains only one Shared project. Renaming is only necessary if a second project with Globbed items is used - in it you need to rename them so that they have different names.

**Reply by jeromelaban** [reply]
Got it! I misunderstood that you were using a second shared project. Indeed, item group names matter in this case.
